### PR TITLE
[Feature] Support tomesd for StableDiffusion speed-up

### PR DESCRIPTION
**Pull request by SheffieldCao**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add `tomesd` support for [Stable Diffusion](https://github.com/open-mmlab/mmagic/blob/main/configs/stable_diffusion/README.md) model and its variants ([ControlNet](https://github.com/open-mmlab/mmagic/blob/main/configs/controlnet/README.md), [DreamBooth](https://github.com/open-mmlab/mmagic/blob/main/configs/dreambooth/README.md)).

## Modification

**Contributions** are mainly as follows:
- [x] Support ToMe patching / unpatching tools `set_tomesd()` and `remove_tomesd`. Refer to [ToMe](https://github.com/dbolya/tomesd).
- [x] Add StableDiffusion inbuilt ToMe patching method `StableDiffusion.set_tomesd()`.
  - [x] `set_tomesd()` for DiffusersWrapper based UNet.
  - [x] `set_tomesd()` for DenoisingUnet in `mmagic`.
- [x] Add using tutorials and some performance evaluation results in [Stable Diffusion](https://github.com/open-mmlab/mmediting/blob/main/configs/stable_diffusion/README.md).
- [x] Add unit test for StableDiffusion `w/ tomesd` and `w/o tomesd`.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmediting/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmediting/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: configs/stable_diffusion/README.md [hunk 1]
```diff
@@ -82,6 +82,122 @@ image = StableDiffuser.infer(prompt)['samples'][0]
...
+    t = time.time()
+    for i in range(100//samples_perprompt):
+        image = sd_model.infer(prompt, height=size, width=size, num_images_per_prompt=samples_perprompt)['samples'][0]
+        if i == 0:
+            image.save(f"{work_dir}/w_tomesd_ratio_{ratio}.png")
+
+    print(f"Generating 100 images with {samples_perprompt} images per prompt, merging ratio {ratio}, time used : {time.time() - t}s")
+```
+
+Here are some performance comparisons:
```
Comment (LeoXing1996) [inline configs/stable_diffusion/README.md:170]: How about including information about the GPU device that was used to evaluate the performance?

File: configs/stable_diffusion/stable-diffusion_ddim_denoisingunet.py [hunk 2]
```diff
@@ -80,4 +80,6 @@
         subfolder='text_encoder'),
     tokenizer=stable_diffusion_v15_url,
     scheduler=diffusion_scheduler,
-    test_scheduler=diffusion_scheduler)
+    test_scheduler=diffusion_scheduler,
+    tomesd_cfg=dict(ratio=0.5),
```
Comment (LeoXing1996) [inline configs/stable_diffusion/stable-diffusion_ddim_denoisingunet.py:84]: I suggest keeping this config unchanged and adding a new config called `stable-diffusion_ddim_denoising-tomesd_0.5.py`.

**Comment by LeoXing1996** [comment]
@SheffieldCao Thanks for this cool feature to speed up stable diffusion sampling! 
This PR looks good enough. It would be great if we could add some simple unit tests to it!

File: configs/stable_diffusion/stable-diffusion_ddim_denoisingunet.py (hunk 2 above)
Comment (SheffieldCao) [reply configs/stable_diffusion/stable-diffusion_ddim_denoisingunet.py]: OK!

**Comment by SheffieldCao** [comment]
Already added unit tests for `tomesd`.

**Comment by SheffieldCao** [comment]
@LeoXing1996 I have checked the `pytest` logs, `ToMe` **ONLY** support torch>=1.11, I add the requirement to `requirements/tests.txt`, but it did not pass the **pr_stage_test/build_cpu**. I'm not sure how to solve the requirements conflicts.

**Comment by SheffieldCao** [comment]
> @SheffieldCao Thanks for this cool feature to speed up stable diffusion sampling! This PR looks good enough. It would be great if we could add some simple unit tests to it!

Already added the unit tests. But `tomesd` requires `torch>=1.12.1` to properly support `torch.Tensor.scatter_reduce`, I just add a pytest skipif to judge whether it is accessible. Also I mentioned this requirement in docs.

**Review by LeoXing1996 — APPROVED** [review]
Great job! Super COOL!

**Pull request merged**